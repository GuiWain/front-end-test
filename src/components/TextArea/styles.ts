import styled, { css, DefaultTheme } from 'styled-components'
import { TextAreaProps } from '.'

export type WrapperProps = {
  error?: boolean
} & Pick<TextAreaProps, 'resize'>

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${TextArea} {
      border-color: ${theme.colors.errorMedium};
    }

    ${Label} {
      color: ${theme.colors.errorMedium};
    }
  `,
  vertical: () => css`
    ${TextArea} {
      resize: vertical;
    }
  `,
  horizontal: () => css`
    ${TextArea} {
      resize: horizontal;
    }
  `,
  both: () => css`
    ${TextArea} {
      resize: both;
    }
  `,
  none: () => css`
    ${TextArea} {
      resize: none;
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, resize }) => css`
    width: 100%;
    margin: 0;
    display: flex;
    gap: 0.5rem;
    flex-direction: column;
    justify-content: center;

    ${!!error && wrapperModifiers.error(theme)}
    ${!!resize && wrapperModifiers[resize]()}
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font: normal normal ${theme.font.semiBold} ${theme.font.sizes.xsmall} / 4rem
      ${theme.font.family};

    letter-spacing: -0.006rem;
    text-align: left;
    color: ${theme.colors.secondaryDark};
    cursor: pointer;
  `}
`

export const TextArea = styled.textarea.attrs({
  'data-testid': 'TextArea'
})`
  ${({ theme }) => css`
    margin: 0;
    font: normal normal ${theme.font.regular} ${theme.font.sizes.xxsmall} /
      2.8rem ${theme.font.family};
    color: ${theme.colors.secondaryDark};

    padding: 1rem;
    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    border: 0.1rem solid ${theme.colors.secondaryLight};
    border-radius: ${theme.border.smallRadius};

    align-self: flex-start;
    width: 100%;

    :placeholder {
      font: normal normal ${theme.font.regular} ${theme.font.sizes.xxsmall} /
        2.8rem ${theme.font.family};
      letter-spacing: 0rem;
      color: ${theme.colors.secondaryLight};
      padding: 2rem;
    }

    :focus {
      outline: none;
      border: 0.2rem solid ${theme.colors.primaryMedium};
      transition: border 0.2s linear;
    }

    &:focus-within {
      box-shadow: 0rem 0.4rem 0.6rem #00000014;
      border: 0.1rem solid ${theme.colors.infoDark};
    }
  `}
`

export const Span = styled.span`
  ${({ theme }) => css`
    display: block;
    font-size: ${theme.font.sizes.xxxsmall};
    color: ${theme.colors.secondaryMedium};
    align-self: center;
  `}
`

export const Error = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.errorMedium};
    font-size: ${theme.font.sizes.xxxsmall};
    align-self: flex-start;
  `}
`
