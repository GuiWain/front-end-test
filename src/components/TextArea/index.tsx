import { useState } from 'react'
import * as S from './styles'

export type TextAreaProps = {
  getValues?: (field: string) => void
  placeholder?: string
  label?: string
  name?: string
  rows?: number
  cols?: number
  maxlength?: number
  defaultValue?: string
  letterCount?: boolean
  resize?: 'vertical' | 'horizontal' | 'both' | 'none'
  error?: string
}

const TextArea = ({
  getValues,
  placeholder,
  label,
  name,
  rows = 5,
  defaultValue,
  error,
  letterCount = false,
  resize = 'none',
  maxlength
}: TextAreaProps) => {
  const [valueLength, setValuelength] = useState(0)

  const onChange = (value: string) => {
    !!getValues && getValues(value)
    letterCount && setValuelength(value.length)
  }

  return (
    <S.Wrapper error={!!error} resize={resize}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.TextArea
        onChange={(e) => onChange(e.currentTarget.value)}
        placeholder={placeholder}
        name={name}
        id={name}
        defaultValue={defaultValue}
        rows={rows}
        maxLength={maxlength}
      />
      {letterCount && (
        <S.Span>
          {valueLength} of {maxlength} caracters
        </S.Span>
      )}
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default TextArea
