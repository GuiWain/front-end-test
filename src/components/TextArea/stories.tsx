import { Story, Meta } from '@storybook/react'
import TextArea, { TextAreaProps } from '.'

export default {
  title: 'Inputs/TextArea',
  component: TextArea,
  argTypes: {
    label: {
      type: 'string'
    },
    placeholder: {
      type: 'string'
    },
    rows: {
      type: 'number'
    },
    maxlength: {
      type: 'number'
    },
    resize: {
      control: {
        type: 'radio',
        options: ['vertical', 'horizontal', 'both', 'none']
      }
    },
    letterCount: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    }
  }
} as Meta

export const Default: Story<TextAreaProps> = (args) => (
  <div style={{ margin: '3rem auto', width: '80%' }}>
    <TextArea {...args} />
  </div>
)

Default.args = {
  label: 'Label label label',
  placeholder: 'Enter your text',
  letterCount: false,
  resize: 'none'
}

export const WithLetterCount: Story<TextAreaProps> = (args) => (
  <div style={{ margin: '3rem auto', width: '80%' }}>
    <TextArea {...args} />
  </div>
)

WithLetterCount.args = {
  label: 'Label label label',
  placeholder: 'Enter your text',
  maxlength: 200,
  letterCount: true,
  resize: 'none'
}

export const WithError: Story<TextAreaProps> = (args) => (
  <div style={{ margin: '3rem auto', width: '80%' }}>
    <TextArea {...args} />
  </div>
)

WithError.args = {
  label: 'Label label label',
  placeholder: 'Enter your text',
  rows: 3,
  error: 'This field is required',
  letterCount: false,
  resize: 'none'
}
