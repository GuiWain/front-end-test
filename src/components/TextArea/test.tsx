import userEvent from '@testing-library/user-event'
import { render, screen } from 'utils/test-utils'

import TextArea from '.'

describe('<TextArea />', () => {
  it('should render the TextArea', () => {
    const { container } = render(<TextArea />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the TextArea with text', () => {
    render(<TextArea />)

    const input = screen.getByTestId('TextArea')
    const text = 'This is my new text'
    userEvent.type(input, text)

    expect(input).toHaveValue(text)
  })

  it('should render the TextArea with Label', () => {
    render(<TextArea label="LabelText" />)

    expect(screen.getByText('LabelText')).toBeInTheDocument()
  })

  it('should render the TextArea with getValues', () => {
    const getValues = jest.fn()
    render(<TextArea getValues={getValues} />)

    const input = screen.getByTestId('TextArea')
    const text = 'This is my new text'
    userEvent.type(input, text)

    expect(getValues).toHaveBeenCalled()
  })

  it('should render the TextArea with error', () => {
    render(<TextArea error="This field is required" />)

    expect(screen.getByText('This field is required')).toBeInTheDocument()
  })

  it('should render the TextArea with resize vertical', () => {
    render(<TextArea resize="vertical" />)

    expect.anything()
  })

  it('should render the TextArea with resize horizontal', () => {
    render(<TextArea resize="horizontal" />)

    expect.anything()
  })

  it('should render the TextArea with resize both', () => {
    render(<TextArea resize="both" />)

    expect.anything()
  })

  it('should render the TextArea with letter count', () => {
    render(<TextArea letterCount={true} maxlength={100} />)

    const input = screen.getByTestId('TextArea')
    const text = 'This is my new text'
    userEvent.type(input, text)

    expect(screen.getByText('19 of 100 caracters')).toBeInTheDocument()
  })
})
