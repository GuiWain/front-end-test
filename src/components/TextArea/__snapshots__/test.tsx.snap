// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should render the TextArea 1`] = `
.c0 {
  width: 100%;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 .c1 {
  resize: none;
}

.c2 {
  margin: 0;
  font: normal normal 400 1.4rem / 2.8rem 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  color: #4F4C4C;
  padding: 1rem;
  background: #FFFFFF 0% 0% no-repeat padding-box;
  border: 0.1rem solid #CCC8C8;
  border-radius: 0.4rem;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  width: 100%;
}

.c2:placeholder {
  font: normal normal 400 1.4rem / 2.8rem 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  color: #CCC8C8;
  padding: 2rem;
}

.c2:focus {
  outline: none;
  border: 0.2rem solid #510161;
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
}

.c2:focus-within {
  box-shadow: 0rem 0.4rem 0.6rem #00000014;
  border: 0.1rem solid #3241A1;
}

<div
  class="c0"
>
  <textarea
    class="c1 c2"
    data-testid="TextArea"
    rows="5"
  />
</div>
`;
